### ISSUE TITLE
Simplify Runner further

### ISSUE THREAD

**Issue by pawelprazak** [pulumi/pulumi-java#668]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Previously in ##642 I've decoupled Stack and Runner and simplified both.

As kindly suggested by @t0yv0 we have an opportunity to simplify Runner even further:
- https://github.com/pulumi/pulumi-java/pull/642/files#r889038759 - try to achieve `<T> CompletableFuture<T> runAsync(Supplier<T> callback);` signature or even `CompletableFuture<Void> run()` if it does not break the semantics.
- we can move the exit code related error handling to the `PulumiInternal` and expose it only in `Pulumi` because it is only useful for the CLI, but not in `PulumiTest` and `PulumiAuto`.

This has implications for the Pulumi, PulumiTest and PulumiAuto APIs (because we need to expose to the user all things needed but not more):
- since the exit code is not useful for the user and is internal-ish, we could stop exposing it in the `Result` (an maybe remove `Result`)
- knowing what errors and exceptions happened is useful for PulumiTest
- knowing what exceptions happened is useful for PulumiAuto, but maybe better semantics would be to just throw on `join`?
- mimicking `pulumi` command can also be useful considerations for better DX: https://github.com/pulumi/pulumi-java/pull/642/files#r888932515

This issue can benefit from more careful consideration regarding consistency in the semantics between different flavours of the API as well as from considering inherent similarities and differences between various modes of operation.

**Comment by t0yv0** [comment]
To clarify, this is an internal refactor with no implications for the user, correct? (It seems to me it is). Thanks!

**Comment by pawelprazak** [comment]
Yes, at its core it's internal refactor, but the content/shape of the internal code needs to be informed by the final API we want, that's why I mentioned it.
